m2cgen/interpreters/haskell/interpreter.py

```diff
@@ -0,0 +1,99 @@
+import os
+
+from m2cgen import ast
+from m2cgen.interpreters import mixins, utils
+from m2cgen.interpreters.haskell.code_generator import HaskellCodeGenerator
+from m2cgen.interpreters.interpreter import ToCodeInterpreter
+
+
+class HaskellInterpreter(ToCodeInterpreter,
+                         mixins.LinearAlgebraMixin):
+    supported_bin_vector_ops = {
+        ast.BinNumOpType.ADD: "addVectors",
+    }
+
+    supported_bin_vector_num_ops = {
+        ast.BinNumOpType.MUL: "mulVectorNumber",
+    }
+
+    exponent_function_name = "exp"
+    sqrt_function_name = "sqrt"
+    tanh_function_name = "tanh"
+
+    def __init__(self,  module_name="Model", indent=4, function_name="score",
+                 *args, **kwargs):
+        self.module_name = module_name
+        self.indent = indent
+        self.function_name = function_name
+
+        cg = HaskellCodeGenerator(indent=indent)
+        super(HaskellInterpreter, self).__init__(cg, *args, **kwargs)
+
+    def interpret(self, expr):
+        self._cg.reset_state()
+        self._reset_reused_expr_cache()
+
+        args = [(True, self._feature_array_name)]
+        func_name = self.function_name
+
+        with self._cg.function_definition(
+                name=func_name,
+                args=args,
+                is_scalar_output=expr.output_size == 1):
+            last_result = self._do_interpret(expr)
+            self._cg.add_code_line(last_result)
+            self._dump_cache()
+
+        if self.with_linear_algebra:
+            filename = os.path.join(
+                os.path.dirname(__file__), "linear_algebra.hs")
+            self._cg.prepend_code_lines(utils.get_file_content(filename))
+
+        self._cg.prepend_code_line(self._cg.tpl_module_definition(
+            module_name=self.module_name))
+
+        return self._cg.code
+
+    def interpret_if_expr(self, expr, if_code_gen=None, **kwargs):
+        if if_code_gen is None:
+            code_gen = HaskellCodeGenerator(indent=self.indent)
+            nested = False
+        else:
+            code_gen = if_code_gen
+            nested = True
+
+        code_gen.add_if_statement(self._do_interpret(
+            expr.test, **kwargs))
+        code_gen.add_code_line(self._do_interpret(
+            expr.body, if_code_gen=code_gen, **kwargs))
+        code_gen.add_else_statement()
+        code_gen.add_code_line(self._do_interpret(
+            expr.orelse, if_code_gen=code_gen, **kwargs))
+        code_gen.add_if_termination()
+
+        if not nested:
+            return self._cache_reused_expr(expr, code_gen.code)
+
+    def interpret_pow_expr(self, expr, **kwargs):
+        base_result = self._do_interpret(expr.base_expr, **kwargs)
+        exp_result = self._do_interpret(expr.exp_expr, **kwargs)
+        return self._cg.infix_expression(
+            left=base_result, right=exp_result, op="**")
+
+    def _cache_reused_expr(self, expr, expr_result):
```

Comment:
It seems like the caching mechanism in Haskell Interpreter is very different from the same mechanism in other languages.

In other languages we cache expression to avoid reevaluation of same expressions more than once. As far as I understand this is not the case here. The only benefit of caching that I'm seeing here is a cleaner generated code where reused expressions are encapsulated into functions. Please correct me if I'm wrong here. If that's the case I'd suggest to leave a brief clarification comment here.

Reply:
On the other hand perhaps GHC is smart enough to not revaluate a function without arguments. Can't quite remember now.

Reply:
> Please correct me if I'm wrong here.

Unfortunately, I can't say anything on this right now. My knowledge of Haskell is limited by two-week course for complete noobs passed in 2017 😃 

I believe, the current implementation is in consistency with the following example:

```
areaTriangleHeron a b c = result           -- use Heron's formula
    where
    result = sqrt (s * (s - a) * (s - b) * (s - c))
    s      = (a + b + c) / 2
```
> https://en.wikibooks.org/wiki/Haskell/Variables_and_functions#Local_definitions

A brief googling gave the following statement:
> GHC does not do any automatic memoization of functions, ...
   https://stackoverflow.com/a/12390953
   https://wiki.haskell.org/Memoization

and the following blog post: https://kseo.github.io/posts/2017-01-14-memoization-in-hasekll.html.

It seems that it's not so trivial to implement the "true cache" mechanism in Haskell. I'm afraid that it is beyond of my current Haskell knowledge and will take much time to learn about it and implement here. But if you have any hints or points for possible implementation, I'll be very grateful and try to improve cache mechanism.

Alternatively, we can live with the current implementation as it's working though not so smart and efficient and create new feature request for improving cache mechanism in Haskell and get back to it later.

Reply:
> On the other hand perhaps GHC is smart enough to not revaluate a function without arguments.

Hmm, I'm afraid to be mistaken, but seems that you are right!

> There is no such thing as a function without parameters (anonymous or otherwise). That's just a value (and in Haskell, actions like main :: IO () are also just values).
  https://stackoverflow.com/a/33211542

> m1 is computed only once because it is a Constant Applicative Form, while m2 is not a CAF, and so is computed for each evaluation.
 https://stackoverflow.com/a/3951131

CAF is a ...

> Any super combinator which is not a lambda abstraction. This includes truly constant expressions such as 12, ((+) 1 2), [1,2,3] as well as partially applied functions such as ((+) 4).
  https://wiki.haskell.org/Constant_applicative_form

Seems that we got "true cache" mechanism with our "zero-argument functions" for free.

Reply:
I found one more way to ensure that Haskell evaluates functions with zero arguments only once: https://stackoverflow.com/a/7721565.

I added `trace "called"` to `func0` to all our snippets from tests. Then compiled them with `-O0` and happily saw `called` only once in all cases.

For example,

```
module Main where
import Debug.Trace

score :: [Double] -> Double
score input =
    (func0) / (func0)
    where
        func0 =
            trace "called"
            exp (1.0)

main = print (score ([1,2,3,4,5]))
```
```
1.0
called
```

Reply:
Awesome, thank you for going such lengths  👍 Now I'll never have this same question again.